community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLog.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.logging;
+
+import org.neo4j.concurrent.AsyncEventSender;
+import org.neo4j.function.Consumer;
+import org.neo4j.logging.AbstractLog;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.Logger;
+
+/**
+ * Wraps a {@link Log}, making its logging asynchronous by letting log statements be treated as events fed into a
+ * {@link AsyncEventSender} to be processed later instead of in the thread making the log statement call.
+ */
+class AsyncLog extends AbstractLog
```

Comment:
The way this is implemented, we still enqueue the logging events, even if the log-level is not enabled.
That is quite wasteful (especially since the finer log-levels are the most likely to not be enabled, and also the most verbose ones).